.github/workflows/ci.yml

```diff
@@ -0,0 +1,330 @@
+name: CI
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+  push:
+    branches:
+      - master
+
+permissions:
+  contents: read 
+
+env: 
+  MSRV: 1.69.0
+
+jobs:
+
+  macos:
+    runs-on: macos-13
+    env: 
+      TARGET: x86_64-apple-darwin
+    steps:
+      - name: checkout
+        uses: actions/checkout@v4
+
+      - name: setup Rust
+        uses: dtolnay/rust-toolchain@master
+        with: 
+          toolchain: '${{ env.MSRV }}'
+          components: clippy
+
+      - name: build
+        uses: ./.github/actions/build
+        with:
+          TARGET: '${{ env.TARGET }}'
+
+      - name: test
+        uses: ./.github/actions/test
+        with:
+          TARGET: '${{ env.TARGET }}'
+
+      - name: before_cache_script
+        run: rm -rf $CARGO_HOME/registry/index
+
+
+  # Use cross for QEMU-based testing
+  # cross needs to execute Docker, GitHub Action already has it installed
+  cross:
+    runs-on: ubuntu-20.04 
+    needs: [rustfmt, minver, macos, linux_native_builds, rust_stable]
+    strategy:
+      fail-fast: true
```

Comment:
`fail-fast` would make GHA cancel all the tasks within a matrix if one of them failed, I think our Cirrus CI does not behave in this way?

Reply:
I think we should set it to false.  "true" means that we could end up building an entire PR several times as the author slowly resolves the various failures in this matrix.  Each iteration would re-build all of the other tasks, so setting fail-fast:true could actually be more work overall than setting it to false.